packages/authentication/src/types.ts

```diff
@@ -22,3 +22,30 @@ export interface UserProfile {
   name?: string;
   email?: string;
 }
+
+/**
+ * An interface that describes the common authentication strategy.
+ *
+ * An authentication strategy is a class with an
+ * 'authenticate' method that verifies a user's credentials and
+ * returns the corresponding user profile.
+ *
+ */
+export interface AuthenticationStrategy {
+  /**
+   * The 'name' property is a unique identifier for the
+   * authentication strategy ( for example : 'basic', 'jwt', etc)
+   */
+  name: string;
+
+  /**
+   * The 'authenticate' method takes in a given request and returns a user profile
+   * which is an instance of 'UserProfile'.
+   * (A user profile is a minimal subset of a user object)
+   * If the user credentials are valid, this method will return a 'UserProfile' instance.
+   * If the user credentials are invalid, this method will throw an error
+   *
```

Comment:
Please explain what is the expected strategy behavior when no credentials were found in the request.

For example:

```
If no user credentials were found or the user credentials are invalid,
this method will throw an error.
```

Reply:
> Please explain what is the expected strategy behavior when no credentials were found in the request.

Hi @bajtos , the authenticate method on a strategy will throw an error like: https://github.com/strongloop/loopback4-example-shopping/blob/master/src/authentication-strategies/JWT.strategy.ts#L34.

Reply:
> > Please explain what is the expected strategy behavior when no credentials were found in the request.
>
> Hi @bajtos , the authenticate method on a strategy will throw an error like: https://github.com/strongloop/loopback4-example-shopping/blob/master/src/authentication-strategies/JWT.strategy.ts#L34.

As I understand that code, a different error will be thrown, see here:
https://github.com/strongloop/loopback4-example-shopping/blob/91805dd8740dce8b19bd7af613c485f7935cc8b5/src/authentication-strategies/JWT.strategy.ts#L21-L22

```ts
    let token = request.query.access_token || request.headers['authorization'];
    if (!token) throw new HttpErrors.Unauthorized('No access token found!');
```

> An error should be thrown by the strategy if the controller method is decorated with an authentication strategy BUT the user has not passed in any credentials

Makes sense (if that's the indented design). I am asking you to capture this information in the tsdoc comment for the new interface.

Reply:
This code is from shopping cart.. it is old code. Going forward the code may be different. The point is that some error is thrown for a given strategy.authenticate().
This PR is about an interface not implementation.